In [1]:
import pandas as pd

historical_data = pd.read_parquet("../data/historical_currency_exchange.parquet")
historical_data["date"] = pd.to_datetime(historical_data["date"])
historical_data

date       AUD       BRL       CAD       EUR       GBP         JPY  \
0  2024-09-30  1.443654  5.434649  1.351196  0.895591  0.747153  142.573268   
1  2024-10-01  1.442917  5.444170  1.352168  0.897557  0.746956  143.831429   
2  2024-10-02  1.449505  5.425444  1.348063  0.903056  0.752241  143.806861   
3  2024-10-03  1.456279  5.442044  1.351451  0.906018  0.754584  146.916773   
4  2024-10-04  1.460930  5.477788  1.355260  0.906452  0.761891  146.592323   
5  2024-10-05  1.471447  5.457431  1.358726  0.910593  0.761939  148.796290   
6  2024-10-06  1.469948  5.463070  1.357479  0.910655  0.762157  148.725322   
7  2024-10-07  1.469358  5.455679  1.357600  0.911148  0.761611  148.509087   
8  2024-10-08  1.479802  5.495973  1.361858  0.910451  0.763520  147.888136   
9  2024-10-09  1.484282  5.535464  1.365741  0.911611  0.763666  148.274271   
10 2024-10-10  1.487595  5.597276  1.371040  0.913740  0.764894  149.176698   
11 2024-10-11  1.484251  5.583092  1.375260  0.914911  0.765915  148.579235   
12 2024-10-12  1.481060  5.611784  1.376385  0.913296  0.765312  149.142242   
13 2024-10-13  1.481516  5.625377  1.376359  0.914137  0.765346  149.163567   
14 2024-10-14  1.484199  5.612327  1.377943  0.915399  0.766102  149.168474   
15 2024-10-15  1.486305  5.596201  1.379823  0.916792  0.765306  149.523126   
16 2024-10-16  1.496513  5.653631  1.378417  0.918318  0.765093  149.025816   
17 2024-10-17  1.492143  5.669771  1.375184  0.920552  0.769540  149.342277   
18 2024-10-18  1.491675  5.653828  1.379385  0.923189  0.768145  149.951370   
19 2024-10-19  1.490596  5.692491  1.380234  0.919512  0.765986  149.542206   
20 2024-10-20  1.490907  5.689223  1.380287  0.920088  0.766265  149.527894   
21 2024-10-21  1.487740  5.637270  1.380375  0.920155  0.766249  149.147331   
22 2024-10-22  1.500639  5.689757  1.383147  0.924220  0.769855  150.683899   
23 2024-10-23  1.498948  5.690853  1.382647  0.926268  0.770607  151.744645   
24 2024-10-24  1.506202  5.689930  1.382894  0.927316  0.773928  152.724433   
25 2024-10-25  1.506748  5.665660  1.384859  0.923876  0.770909  151.903627   
26 2024-10-26  1.513840  5.708040  1.389294  0.925752  0.771547  152.268731   
27 2024-10-27  1.514088  5.707994  1.389221  0.926135  0.771502  152.283657   
28 2024-10-28  1.515760  5.708373  1.389687  0.926519  0.771935  153.800613   
29 2024-10-29  1.518902  5.711547  1.388779  0.924510  0.770736  152.951742   
30 2024-10-30  1.522229  5.761654  1.390965  0.923864  0.768083  153.160003   

          MXN          NGN        RUB       SGD  USD  
0   19.694724  1668.736400  94.133735  1.280156    1  
1   19.667561  1670.694524  92.898519  1.284352    1  
2   19.606748  1669.653006  94.583198  1.286983    1  
3   19.457701  1670.097873  95.655442  1.294391    1  
4   19.363467  1649.763738  94.755337  1.296800    1  
5   19.286645  1619.465727  94.934698  1.303670    1  
6   19.284784  1626.760323  94.993467  1.304541    1  
7   19.263497  1658.088878  95.582401  1.303118    1  
8   19.297377  1626.621519  96.167634  1.302788    1  
9   19.357738  1620.528912  96.750546  1.304275    1  
10  19.465534  1620.288429  97.076400  1.306731    1  
11  19.457569  1620.553279  96.608874  1.306209    1  
12  19.287951  1639.315783  95.683504  1.305143    1  
13  19.277162  1619.450022  95.818212  1.304986    1  
14  19.301100  1620.334223  96.250849  1.305672    1  
15  19.416737  1635.134961  95.165613  1.308048    1  
16  19.711436  1635.438782  96.939582  1.308750    1  
17  19.895284  1635.903564  97.330418  1.311236    1  
18  19.819733  1636.731886  97.211085  1.314172    1  
19  19.880298  1645.268713  95.880596  1.310179    1  
20  19.883334  1637.845302  95.319348  1.310248    1  
21  19.892753  1634.657490  95.891940  1.309439    1  
22  19.957118  1642.553927  96.798098  1.315318    1  
23  19.967148  1644.825044  95.749323  1.317586    1  
24  19.831042  1642.896306  95.922769  1.321859    1  
25  19.824089  1642.861102  96.245440  1.318615   

In [2]:
fraud_data = pd.read_parquet("../data/transaction_fraud_data.parquet")
fraud_data['timestamp'] = pd.to_datetime(fraud_data['timestamp'])
fraud_data['date'] = pd.to_datetime(fraud_data['timestamp'].dt.date)
hist_long = (
    historical_data
    .melt(id_vars=['date'], var_name='currency', value_name='rate')
)
fraud_data = fraud_data.merge(
    hist_long,
    on=['date', 'currency'],
    how='left'
)
fraud_data['amount_usd'] = fraud_data['amount'] / fraud_data['rate']
fraud_data

transaction_id customer_id       card_number  \
0          TX_a0ad2a2a  CUST_72886  6646734767813109   
1          TX_3599c101  CUST_70474   376800864692727   
2          TX_a9461c6d  CUST_10715  5251909460951913   
3          TX_7be21fc4  CUST_16193   376079286931183   
4          TX_150f490b  CUST_87572  6172948052178810   
...                ...         ...               ...   
7483761    TX_58a762fa  CUST_11344   377055034743699   
7483762    TX_dabcf671  CUST_38961   373625915641065   
7483763    TX_ab567eb4  CUST_75576  6702342842796660   
7483764    TX_5ae61846  CUST_82508   374615717295581   
7483765    TX_54fccca9  CUST_20139   373017396639208   

                         timestamp vendor_category vendor_type  \
0       2024-09-30 00:00:01.034820      Restaurant   fast_food   
1       2024-09-30 00:00:01.764464   Entertainment      gaming   
2       2024-09-30 00:00:02.273762         Grocery    physical   
3       2024-09-30 00:00:02.297466             Gas       major   
4       2024-09-30 00:00:02.544063      Healthcare     medical   
...                            ...             ...         ...   
7483761 2024-10-30 23:59:58.926575          Retail      online   
7483762 2024-10-30 23:59:58.950801         Grocery    physical   
7483763 2024-10-30 23:59:58.972155         Grocery    physical   
7483764 2024-10-30 23:59:58.996608          Retail      online   
7483765 2024-10-30 23:59:59.101885         Grocery      online   

                  vendor     amount currency    country  ...  \
0              Taco Bell     294.87      GBP         UK  ...   
1                  Steam    3368.97      BRL     Brazil  ...   
2            Whole Foods  102582.38      JPY      Japan  ...   
3                  Exxon     630.60      AUD  Australia  ...   
4         Medical Center  724949.27      NGN    Nigeria  ...   
...                  ...        ...      ...        ...  ...   
7483761       AliExpress     887.32      SGD  Singapore  ...   
7483762           Kroger     356.06      EUR     France  ...   
7483763           Kroger     391.96      SGD  Singapore  ...   
7483764             eBay     601.71      USD        USA  ...   
7483765  Walmart Grocery   40668.06      JPY      Japan  ...   

                       device_fingerprint       ip_address  \
0        e8e6160445c935fd0001501e4cbac8bc   197.153.60.199   
1        a73043a57091e775af37f252b3a32af9  208.123.221.203   
2        218864e94ceaa41577d216b149722261   10.194.159.204   
3        70423fa3a1e74d01203cf93b51b9631d   17.230.177.225   
4        9880776c7b6038f2af86bd4e18a1b1a4  136.241.219.151   
...                                   ...              ...   
7483761  2e37bb086e754bfaf2ec3806dcac8d98    248.8.111.152   
7483762  459c67dbfbb15019df8088fc0f2ffe1d   224.123.137.90   
7483763  337d36a8f5e811564a1dd3c255db01dd     31.196.9.234   
7483764  52874c17f4ab0846fac1aed0c66e8551     188.74.107.8   
7483765  e9e0e0db1c95f5f44613492b2f931fac    205.233.99.48   

        is_outside_home_country  is_high_risk_vendor is_weekend  \
0                         False                False      False   
1                          True                 True      False   
2                         False                False      False   
3                         False                False      False   
4                          True                False      False   
...                         ...                  ...        ...   
7483761                   False                False      False   
7483762                   False                False      False   
7483763                   False                False      False   
7483764                   False                False      False   
7483765                   False                False      False   

                                        last_hour_activity is_fraud  \
0        {'num_transactions': 1197, 'total_amount': 334...    False   
1        {'num_transactions': 509, 'total_amount': 2011...     True   
2        {

In [3]:
bool_cols = [
    "is_card_present",
    "is_outside_home_country",
    "is_high_risk_vendor",
    "is_weekend",
    "is_fraud",
]
fraud_data[bool_cols] = fraud_data[bool_cols].astype(bool)

In [4]:
fraud_data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 7483766 entries, 0 to 7483765
Data columns (total 26 columns):
 #   Column                   Dtype         
---  ------                   -----         
 0   transaction_id           object        
 1   customer_id              object        
 2   card_number              int64         
 3   timestamp                datetime64[us]
 4   vendor_category          object        
 5   vendor_type              object        
 6   vendor                   object        
 7   amount                   float64       
 8   currency                 object        
 9   country                  object        
 10  city                     object        
 11  city_size                object        
 12  card_type                object        
 13  is_card_present          bool          
 14  device                   object        
 15  channel                  object        
 16  device_fingerprint       object        
 17  ip_address               ob

In [5]:
ts = pd.to_datetime(fraud_data["timestamp"], errors="coerce")
fraud_data["hour"] = ts.dt.hour
fraud_data["dow"] = ts.dt.dayofweek  # 0=Mon
fraud_data["month"] = ts.dt.month
fraud_data["is_night"] = fraud_data["hour"].isin([0, 1, 2, 3, 4, 5]).astype(int)

In [6]:
import numpy as np

fraud_data["amount_log1p"] = np.log1p(fraud_data["amount_usd"].clip(lower=0))

In [7]:
fraud_data["city_unknown"] = (
    fraud_data["city"].str.contains("Unknown", case=False, na=False).astype(int)
)
fraud_data["vendor_unknown"] = fraud_data["vendor"].isna().astype(int)
fraud_data["currency"] = fraud_data["currency"].fillna("UNK")
fraud_data["vendor_category"] = fraud_data["vendor_category"].fillna("UNK")
fraud_data["vendor_type"] = fraud_data["vendor_type"].fillna("UNK")
fraud_data["device"] = fraud_data["device"].fillna("UNK")
fraud_data["channel"] = fraud_data["channel"].fillna("UNK")
fraud_data["card_type"] = fraud_data["card_type"].fillna("UNK")
fraud_data["city_size"] = fraud_data["city_size"].fillna("UNK")
fraud_data["country"] = fraud_data["country"].fillna("UNK")

In [8]:
low_card_cats = [
    "vendor_category",
    "vendor_type",
    "card_type",
    "device",
    "channel",
    "city_size",
    "currency",
]
K = 20
for c in low_card_cats:
    topk = fraud_data[c].value_counts().nlargest(K).index
    fraud_data[c] = np.where(fraud_data[c].isin(topk), fraud_data[c], "OTHER")

In [9]:
bin_cols = [
    "is_card_present",
    "is_outside_home_country",
    "is_high_risk_vendor",
    "is_weekend",
    "is_night",
    "city_unknown",
    "vendor_unknown",
]
for c in bin_cols:
    fraud_data[c] = fraud_data[c].astype(int)

In [10]:
num_cols = ["amount_log1p", "hour", "dow", "month"]
cat_cols = low_card_cats + []
bin_cols = bin_cols + []

In [11]:
X_num = fraud_data[num_cols].astype("float32")

In [12]:
from pandas import DataFrame


def one_hot_simple(series):
    # explicit, to avoid hidden placeholders
    vals = series.astype("category")
    mat = pd.get_dummies(vals, prefix=series.name, dtype=np.uint8)
    return mat


X_cat_list = [one_hot_simple(fraud_data[c]) for c in cat_cols]
X_cat = (
    pd.concat(X_cat_list, axis=1).astype("uint8")
    if X_cat_list
    else pd.DataFrame(index=fraud_data.index)
)

X_bin = fraud_data[bin_cols].astype("uint8")

# Concatenate final matrix
X: DataFrame = pd.concat([X_num, X_bin, X_cat], axis=1)
y = fraud_data["is_fraud"].astype(int)  # only for validation/insight

In [ ]:
from pyod.models.ecod import ECOD
from pyod.models.iforest import IForest
from sklearn.preprocessing import StandardScaler
from tqdm import tqdm

X_model = X.astype("float32").values

scaler = StandardScaler(with_mean=True, with_std=True)
X_scaled = scaler.fit_transform(X_model)

ecod = ECOD(n_jobs=-1)  # fast, good baselines
ifo = IForest(n_estimators=200, max_samples="auto", random_state=42, n_jobs=-1)

models = [
    ecod,
    ifo,
]
for m in tqdm(models):
    m.fit(X_scaled)


  0%|          | 0/2 [00:00<?, ?it/s][Parallel(n_jobs=8)]: Using backend LokyBackend with 8 concurrent workers.
[Parallel(n_jobs=8)]: Done   2 out of   8 | elapsed:   35.5s remaining:  1.8min
[Parallel(n_jobs=8)]: Done   8 out of   8 | elapsed:   43.9s finished
100%|██████████| 2/2 [03:58<00:00, 119.07s/it]


In [14]:
# Collect raw scores and z-score standardize each model’s output before ensembling
def zscore(a):
    a = np.asarray(a, dtype=np.float64)
    return (a - np.nanmean(a)) / (np.nanstd(a) + 1e-12)


scores = np.column_stack([zscore(m.decision_scores_) for m in models])
score_avg = scores.mean(axis=1)
score_max = scores.max(axis=1)

# Attach back to frame
fraud_data["anomaly_score_avg"] = score_avg.astype("float32")
fraud_data["anomaly_score_max"] = score_max.astype("float32")

In [15]:
fraud_data

transaction_id customer_id       card_number  \
0          TX_a0ad2a2a  CUST_72886  6646734767813109   
1          TX_3599c101  CUST_70474   376800864692727   
2          TX_a9461c6d  CUST_10715  5251909460951913   
3          TX_7be21fc4  CUST_16193   376079286931183   
4          TX_150f490b  CUST_87572  6172948052178810   
...                ...         ...               ...   
7483761    TX_58a762fa  CUST_11344   377055034743699   
7483762    TX_dabcf671  CUST_38961   373625915641065   
7483763    TX_ab567eb4  CUST_75576  6702342842796660   
7483764    TX_5ae61846  CUST_82508   374615717295581   
7483765    TX_54fccca9  CUST_20139   373017396639208   

                         timestamp vendor_category vendor_type  \
0       2024-09-30 00:00:01.034820      Restaurant   fast_food   
1       2024-09-30 00:00:01.764464   Entertainment      gaming   
2       2024-09-30 00:00:02.273762         Grocery    physical   
3       2024-09-30 00:00:02.297466             Gas       major   
4       2024-09-30 00:00:02.544063      Healthcare     medical   
...                            ...             ...         ...   
7483761 2024-10-30 23:59:58.926575          Retail      online   
7483762 2024-10-30 23:59:58.950801         Grocery    physical   
7483763 2024-10-30 23:59:58.972155         Grocery    physical   
7483764 2024-10-30 23:59:58.996608          Retail      online   
7483765 2024-10-30 23:59:59.101885         Grocery      online   

                  vendor     amount currency    country  ...  amount_usd hour  \
0              Taco Bell     294.87      GBP         UK  ...  394.658011    0   
1                  Steam    3368.97      BRL     Brazil  ...  619.905757    0   
2            Whole Foods  102582.38      JPY      Japan  ...  719.506409    0   
3                  Exxon     630.60      AUD  Australia  ...  436.808287    0   
4         Medical Center  724949.27      NGN    Nigeria  ...  434.430069    0   
...                  ...        ...      ...        ...  ...         ...  ...   
7483761       AliExpress     887.32      SGD  Singapore  ...  670.362079   23   
7483762           Kroger     356.06      EUR     France  ...  385.402905   23   
7483763           Kroger     391.96      SGD  Singapore  ...  296.122166   23   
7483764             eBay     601.71      USD        USA  ...  601.710000   23   
7483765  Walmart Grocery   40668.06      JPY      Japan  ...  265.526634   23   

        dow  month is_night amount_log1p city_unknown vendor_unknown  \
0         0      9        1     5.980550            1              0   
1         0      9        1     6.431179            1              0   
2         0      9        1     6.579954            1              0   
3         0      9        1     6.081781            1              0   
4         0      9        1     6.076334            1              0   
...      ..    ...      ...          ...          ...            ...   
7483761   2     10        0     6.509309            1              0   
7483762   2     10        0     5.956881            1              0   
7483763   2     10        0     5.694143            1              0   
7483764   2     10        0     6.401436            0              0   
7483765   2     10        0     5.585474            1              0   

         anomaly_score_avg  anomaly_score_max  
0                 1.684935           1.691389  
1                 1.867401           1.869367  
2                 1.023649           1.249033  
3                 1.500546           1.638044  
4                 1.566811           1.794430  
...                    ...                ...  
7483761          -0.819483          -0.644906  
7483762          -0.969088          -0.879803  
7483763          -0.691573          -0.672286  
7483764           1.240138           1.999833  
7483765          -0.941622          -0.723134  

[7483766 rows x 35 columns]

In [16]:
from sklearn.metrics import precision_recall_fscore_support, average_precision_score
from pprint import pprint

fraud_rate = float(y.mean())  # e.g., 0.002 (0.2%) — compute from your data
quantile = 1.0 - fraud_rate if fraud_rate > 0 and fraud_rate < 1 else 0.995  # fallback
th = fraud_data["anomaly_score_avg"].quantile(quantile)

fraud_data["anomaly_flag"] = (fraud_data["anomaly_score_avg"] >= th).astype(int)

precision, recall, f1, _ = precision_recall_fscore_support(
    fraud_data["is_fraud"],
    fraud_data["anomaly_flag"],
    average="binary",
    zero_division=0,
)
ap = average_precision_score(fraud_data["is_fraud"], fraud_data["anomaly_score_avg"])

metrics = {
    "fraud_rate": fraud_rate,
    "anomaly_rate": float(fraud_data["anomaly_flag"].mean()),
    "precision_at_calibrated_threshold": float(precision),
    "recall_at_calibrated_threshold": float(recall),
    "f1_at_calibrated_threshold": float(f1),
    "avg_precision_PR_AUC": float(ap),
}
pprint(metrics)


{'anomaly_rate': 0.19972818498066347,
 'avg_precision_PR_AUC': 0.6582814825987734,
 'f1_at_calibrated_threshold': 0.6201433179079144,
 'fraud_rate': 0.19972818498066347,
 'precision_at_calibrated_threshold': 0.6201433179079144,
 'recall_at_calibrated_threshold': 0.6201433179079144}


**Ключевые выводы:**

- Уровень аномалий (≈19.97 %) совпадает с фактическим уровнем мошенничества, что свидетельствует об адекватной калибровке порога.  
- Precision и Recall на выбранном пороге равны ≈0.62, что указывает на сбалансированную детекцию мошеннических транзакций.  
- Среднее значение Precision-Recall AUC ≈0.66 демонстрирует хорошую способность модели отличать нормальные операции от аномалий.  

In [19]:
# Helper: summarize by slice with clear KPIs
def slice_report(
    frame,
    by,
    score="anomaly_score_avg",
    flag="anomaly_flag",
    label="is_fraud",
    min_tx=500,
):
    g = (
        frame.groupby(by)
        .agg(
            tx=("transaction_id", "count"),
            anomaly_rate=(flag, "mean"),
            fraud_rate=(label, "mean"),
            score_p95=(score, lambda s: np.quantile(s, 0.95)),
            score_mean=(score, "mean"),
            amount_mean=("amount_usd", "mean"),
            amount_p95=("amount_usd", lambda s: np.quantile(s, 0.95)),
            outside_home_rate=("is_outside_home_country", "mean"),
            high_risk_vendor_rate=("is_high_risk_vendor", "mean"),
            card_present_rate=("is_card_present", "mean"),
            weekend_rate=("is_weekend", "mean"),
        )
        .reset_index()
    )
    g = g[g["tx"] >= min_tx].copy()
    baseline_anom = frame[flag].mean()
    baseline_fraud = frame[label].mean()
    g["anomaly_lift"] = g["anomaly_rate"] / max(baseline_anom, 1e-9)
    g["fraud_lift"] = (
        g["fraud_rate"] / max(baseline_fraud, 1e-9) if baseline_fraud > 0 else np.nan
    )
    g = g.sort_values(
        ["anomaly_lift", "score_p95", "tx"], ascending=[False, False, False]
    )
    return g


by_lists = {
    "vendor_category": ["vendor_category"],
    "channel": ["channel"],
    "device": ["device"],
    "country": ["country"],
    "card_type": ["card_type"],
    "hour_of_day": ["hour"],
    "vendor_type": ["vendor_type"],
    "city_size": ["city_size"],
    "country_vendorcat": ["country", "vendor_category"],
}

reports = {
    name: slice_report(fraud_data, by, min_tx=1000) for name, by in by_lists.items()
}

# Top hotspots examples
for name, rep in reports.items():
    print(f"\n=== Top hotspots: {name} ===")
    print(rep.head(10).to_string(index=False))



=== Top hotspots: vendor_category ===
vendor_category     tx  anomaly_rate  fraud_rate  score_p95  score_mean  amount_mean  amount_p95  outside_home_rate  high_risk_vendor_rate  card_present_rate  weekend_rate  anomaly_lift  fraud_lift
         Travel 935790      0.329088    0.200341   2.705400    0.504948  1040.826006 3203.264128           0.323227                    1.0           0.087380      0.257768      1.647678    1.003068
  Entertainment 936173      0.274566    0.199632   2.443148    0.364502   350.314504 1249.823204           0.322088                    1.0           0.086937      0.258101      1.374697    0.999518
     Restaurant 936178      0.198708    0.199696   2.175941    0.076974   323.328116 1237.540051           0.322081                    0.0           0.086816      0.258006      0.994892    0.999839
     Healthcare 936770      0.183396    0.199376   2.129787   -0.001411   514.887331 1629.225241           0.321798                    0.0           0.086975      0.2583

### Краткий анализ топ-хотспотов

- **Категории товаров**  
  - Travel: anomaly_rate ≈ 33 % (anomaly_lift 1.65) — максимальный уровень подозрительных транзакций.  
  - Entertainment: anomaly_rate ≈ 27 % (anomaly_lift 1.37).  
  - Restaurant, Healthcare, Gas: anomaly_rate близки к среднему (~16–20 %).

- **Каналы оплаты**  
  - POS: anomaly_rate ≈ 100 % (anomaly_lift 5.01) и fraud_rate 100 % — все POS-транзакции попали в топ.  
  - Mobile: anomaly_rate ≈ 21 % (anomaly_lift 1.05).  
  - Web: anomaly_rate ≈ 8 % (anomaly_lift 0.41).

- **Устройства**  
  - Chip Reader / NFC / Magnetic Stripe: anomaly_rate ≈ 100 % (anomaly_lift 5.01).  
  - Android App: anomaly_rate ≈ 23.5 % (anomaly_lift 1.18).  
  - iOS App: anomaly_rate ≈ 18.3 % (anomaly_lift 0.92).

- **Страны**  
  - USA: anomaly_rate ≈ 59 % (anomaly_lift 2.97), но низкий fraud_rate ≈ 7 % (fraud_lift 0.37).  
  - Mexico / Brazil / Russia: anomaly_rate ≈ 26–27 % (anomaly_lift 1.29–1.36), fraud_lift ≈ 1.85–1.90.

- **Тип карты**  
  - Platinum Credit: anomaly_rate ≈ 21 % (anomaly_lift 1.05).  
  - Остальные типы карт — около среднего (≈ 19 %).

- **Время суток**  
  - Часы 1–4 AM: anomaly_rate ≈ 49–59 % (anomaly_lift 2.44–2.94), fraud_lift ≈ 2.96.  
  - Часы 5–6 AM: anomaly_rate ≈ 17–29 % (anomaly_lift 1.00–1.47).

- **Тип поставщика (vendor_type)**  
  - transport, airlines, booking: anomaly_rate ≈ 35–37 % (anomaly_lift 1.73–1.84).  
  - streaming / gaming: anomaly_rate ≈ 28–30 % (anomaly_lift 1.42–1.48).

- **Размер города**  
  - large: anomaly_rate ≈ 100 % (anomaly_lift 5.01), fraud_rate низкий (≈ 7 %).  
  - medium: anomaly_rate ≈ 18 % (anomaly_lift 0.89).

- **Комбинация «страна × категория»**  
  - USA×Travel: anomaly_rate ≈ 81 % (anomaly_lift 4.04), fraud_lift 0.38.  
  - Mexico×Travel: anomaly_rate ≈ 41 % (anomaly_lift 2.05), fraud_lift 1.91.  


In [20]:
# Top anomalous confirmed frauds (for pattern reverse-engineering)
fraud_top = fraud_data[fraud_data["is_fraud"] == 1].sort_values(
    "anomaly_score_avg", ascending=False
)
print("\nTop 20 fraud transactions by anomaly score:\n")
print(
    fraud_top[
        [
            "transaction_id",
            "customer_id",
            "vendor",
            "vendor_category",
            "channel",
            "country",
            "amount_usd",
            "anomaly_score_avg",
        ]
    ]
    .head(20)
    .to_string(index=False)
)

# Top anomalous non-frauds (for false-positive analysis and rule refinement)
nonfraud_top = fraud_data[fraud_data["is_fraud"] == 0].sort_values(
    "anomaly_score_avg", ascending=False
)
print("\nTop 20 non-fraud anomalies (candidates for policy tuning):\n")
print(
    nonfraud_top[
        [
            "transaction_id",
            "customer_id",
            "vendor",
            "vendor_category",
            "channel",
            "country",
            "amount_usd",
            "anomaly_score_avg",
        ]
    ]
    .head(20)
    .to_string(index=False)
)

# Per-customer risk snapshots (useful to spot mule/compromised accounts)
cust_view = (
    fraud_data.groupby("customer_id")
    .agg(
        tx=("transaction_id", "count"),
        anomalies=("anomaly_flag", "sum"),
        anomaly_rate=("anomaly_flag", "mean"),
        frauds=("is_fraud", "sum"),
        total_amount=("amount_usd", "sum"),
        max_score=("anomaly_score_avg", "max"),
    )
    .sort_values(["anomalies", "max_score", "tx"], ascending=[False, False, False])
)

print("\nCustomers with the most anomalies:\n")
print(cust_view.head(20).to_string())



Top 20 fraud transactions by anomaly score:

transaction_id customer_id            vendor vendor_category channel country  amount_usd  anomaly_score_avg
   TX_aa306af1  CUST_78080             Delta          Travel     pos     USA     2967.12           6.448431
   TX_3963a045  CUST_40700            Airbnb          Travel     pos     USA    14794.52           6.434097
   TX_3ba114b2  CUST_27724             Delta          Travel     pos     USA        4.41           6.318048
   TX_79da512a  CUST_29582        Epic Games   Entertainment     pos     USA        1.48           6.295519
   TX_2f1bde67  CUST_56637             Hyatt          Travel     pos     USA     4256.20           6.254673
   TX_b6050687  CUST_24094   United Airlines          Travel     pos     USA     7049.06           6.253131
   TX_9a1d4cdd  CUST_76325         Southwest          Travel     pos     USA     3633.56           6.198495
   TX_81071316  CUST_38543        Hotels.com          Travel     pos     USA     5690.99  

**Краткий анализ результатов**

- **Топ-20 мошеннических транзакций:**
  - Абсолютное большинство приходится на категорию **Travel** (авиаперевозки, отели, бронирование) через **POS** в США.
  - Высокие суммы (до \$14 000+) сочетаются с экстремальными скоринговыми значениями (≈5.8–6.4), что говорит об однозначно подозрительных операциях.
  - Низкие суммы (минимум \$1–4) также получают высокие оценки: вероятно, мошенники путают маленькие тестовые транзакции перед крупными расходами.

- **Топ-20 аномалий без факта мошенничества:**
  - Большинство — мобильные **Travel**-транзакции в США (Delta, Booking.com, Kayak и пр.).
  - Скоры в диапазоне 4.4–4.9 указывают на завышенный риск-профиль: возможно, стоит пересмотреть пороговые значения для мобильного канала.
  - Появляются и другие категории (Restaurant, Entertainment), что может указывать на ложные срабатывания при необычных, но легитимных платежах.

- **Клиенты с наибольшим числом аномалий:**
  - Топ-20 клиентов имеют от ≈1 800 до ≈3 700 транзакций и аномалию в 34–66 % операций.
  - У нескольких клиентов (например, CUST_99677, CUST_35081) аномалий даже больше, чем у мошенников (anomaly_rate > fraud_rate) — кандидаты на углублённый анализ и проверку.
  - Высокие суммарные обороты (до \$3 000 000) сочетаются с экстремальными скоринговыми пиками (>5.3) — приоритизировать расследование этих аккаунтов.


In [21]:
# Numeric drivers: rank by correlation with anomaly score
num_drivers = (
    fraud_data[num_cols + ['anomaly_score_avg']]
    .corr(numeric_only=True)['anomaly_score_avg']
    .drop('anomaly_score_avg').sort_values(ascending=False)
)
print("\nNumeric drivers of anomaly score (corr):\n", num_drivers.head(15).to_string())

# Categorical drivers: mean score by level vs overall
overall = fraud_data['anomaly_score_avg'].mean()
for c in ['vendor_category','channel','device','country','card_type','city_size']:
    means = fraud_data.groupby(c)['anomaly_score_avg'].mean().sort_values(ascending=False)
    lift = (means / overall).rename('score_lift').to_frame().join(
        fraud_data.groupby(c)['is_fraud'].mean().rename('fraud_rate')
    )
    print(f"\nScore & fraud lift by {c} (top 10):\n")
    print(lift.head(10).to_string())



Numeric drivers of anomaly score (corr):
 dow             0.087674
month          -0.175177
amount_log1p   -0.184976
hour           -0.283986

Score & fraud lift by vendor_category (top 10):

                   score_lift  fraud_rate
vendor_category                          
Travel           1.161609e+08    0.200341
Entertainment    8.385180e+07    0.199632
Restaurant       1.770751e+07    0.199696
Healthcare      -3.245524e+05    0.199376
Gas             -3.113483e+07    0.199731
Education       -4.982261e+07    0.199459
Grocery         -5.849898e+07    0.200194
Retail          -7.821551e+07    0.199398

Score & fraud lift by channel (top 10):

          score_lift  fraud_rate
channel                         
pos      493006560.0    1.000000
mobile    31668446.0    0.123878
web      -86090792.0    0.123275

Score & fraud lift by device (top 10):

                  score_lift  fraud_rate
device                                  
Chip Reader      508837216.0    1.000000
NFC Payment     

**Числовые драйверы аномалий:**
- `hour` (corr ≈ −0.284): с ростом часа суток средний anomaly_score падает — ночные часы более риск-оформлены.  
- `amount_log1p` (corr ≈ −0.185): крупные суммы чуть менее “аномальны” в среднем, возможно из-за нормальных крупных бизнес-транзакций.  
- `month` (corr ≈ −0.175): в поздние месяцы года чуть ниже средний риск.  
- `dow` (corr ≈ +0.088): небольшой рост риска в выходные дни.

---

**Категориальные драйверы (score lift):**
- **vendor_category**:  
  - **Travel** и **Entertainment** дают гигантский uplift (≈ 1e8), остальные категории — отрицательный или близкий к нулю.  
- **channel**:  
  - **POS** — экстремальный lift (≈ 4.9e8), **mobile** положительный (≈ 3.2e7), **web** — отрицательный.  
- **device**:  
  - Чип-ридеры, NFC и магнитная полоса — очень высокий uplift (> 4.8e8).  
  - Мобильные приложения (Android/iOS) — умеренный позитив (≈ 4.6e7 / 1.8e7).  
  - Браузеры (Safari/Firefox/Chrome/Edge) — отрицательный lift.  
- **country**:  
  - **USA**, **Mexico**, **Brazil**, **Nigeria**, **Russia** — положительный uplift; **Singapore**, **UK**, **Australia**, **Canada**, **Japan** — отрицательный.  
- **card_type**:  
  - **Platinum Credit** — единственный значимый позитив (≈ 1.5e7), остальные типы карт имеют отрицательный uplift.  
- **city_size**:  
  - **large** — очень высокий uplift (≈ 4.7e8), **medium** — отрицательный.  
